README.md
# Book-marker
Book marker script


CHANGES
# ---------------------------------------------------------------
# bookmarker
# A WWW based bookmark management, retrieval, and search tool.
# Copyright (C) 1998  Padraic Renaghan
# Licensed under terms of GNU General Public License
# (see http://www.renaghan.com/bookmarker/source/LICENSE)
# ---------------------------------------------------------------

11/01/2000   Version 2.8.0
 - changed comments in pgsql create file to -- as
   suggested by jlc6@po.cwru.edu
 - changed public to public_f in bkshared.inc
 - fixed handling of user_total_bookmarks.
   NOTE: if upgrading, make sure and delete all rows
   in active_sessions table to allow this fix to work!
   (thanks to gburca@yahoo.com)
 - fixed sort/control-break logic in list pages to use
   category/subcat names, not IDs to support viewing
   bookmarks from multiple users
   (thanks to yamamoto@trustbee.com)
 - fixed delete logic in codes.php3 to correctly update
   bookmark table
   (thanks to yamamoto@trustbee.com)
       
08/10/2000   Version 2.7.0
 - removed guest user settings. now using PHPLIB default
   authentication to handle public access.

07/17/2000   Version 2.6.0
 - added escape() to quik-mark and mail-this-link javascript to 
   handle ampersands properly
   you'll want to update your quik-mark and mail-this-link bookmarks
   in your browser!
 - fixed refresh/cache issues with MSIE and Netscape on the tree
   view page (I hope!). It still is not 100% correct. It all works
   better if you are logged in before you attempt to go to the tree
   view page.
 - fixed quote handling on search page as suggested by
   seva@rubis.lod.com
 - added ODBC/MS Sql Server support as patched by 
   seva@rubis.lod.com

07/06/2000   Version 2.5.0
 - changed default code names from "unassigned" and "not rated"
   to "--"
 - added links to edit codes from maintain and create pages.
 - enhanced import page to look for existing category and
   subcategory before adding a new one. This will allow you
   to import mutliple files/times without getting duplicate
   categories or subcategories. Note the duplicate check for
   bookmarks was already coded.

06/29/2000   Version 2.4.0
 - changed from FastTemplates to PHPLIB's template class
 - changed method by which db callout is instantiated
   to make installation easier. Encouraged by idea from
   javaman@vnet.net

06/13/2000   Version 2.3.0
 - fixed column name bug, uid, in user.php3
 - changed to not use include_path anymore since it
   causes much confusion and problems for people.
   Using define() globals now.
 - from jeferee@pobox.com: user page now adds ratings
   and uses nextid function for bookmark insert.
 - removed reliance on .htaccess file for settings.
   trying to make installation and use easier for the
   not technie user. added checks to index page to test
   for proper php settings.

04/12/yyyy   Version 2.2.0
 - upgraded to PHPLIB 7.2b which is supposedly PHP4 compat.
   small change required to auth_user for this PHPLIB upgrade,
   see alter-2.2.0.sql.
 - applied PHP4 related changes from <toby.cabot@pobox.com>
 - fixed maillink sql bug identified by <toby.cabot@pobox.com>
 - changed all auto_increment db columns to use PHPLIB db_sequence
   and nextid functionality. this makes the db code more generic.
 - added note to import page about MSIE export feature.
 - first attempt at PostgreSQL support by adding database callout
   class and functions to separate database specific code from
   the main code base.
 - moved all alter and db create scripts to ./lib/db/

03/24/2000   Version 2.1.0
 - changed name of column bookmark.public to bookmark.public_f
   to better support other databases like postgres.
 - to bookmarker.inc added db_first_row_offset to support use
   of OFFSET, LIMIT SQL in other databases like postgres.
 - import now lists all folder levels from input file in the
   keyword column to facilitate searching since only 2 folder
   levels are directly supported in the db.
 - added WAP/WML sample page that displays public WML bookmarks
 - when a cat/subcat is deleted, bookmarks of the cat/subcat
   are assigned to cat/subcat 0.
 - disallow delete of cat/subcat 0 to support above.
 - fix wording of cat/subcat/rating section to refer to user
   preferences page
 - add BIG_TABLES db option to plist before main query.
   this is MySQL specific, sorry.
 - logout link only shows when user is logged in
 - logout process redirects user back to start page - no need
   for logout.body.tpl any longer


01/23/2000   Version 2.0.2
 - fixed bad auth_user update sql in alter-2.0.0.sql and bookmarker.sql.
   this was causing more than one auth_user row to have the same 
   perm_auth_cookie value, which could enable someone to login as
   another user. See UPGRADE for info on how to fix this.
 - changed composition of cookie to include user ID and be base_64
   encoded before it is sent to the browser.
 - added perm_cookie_lifetime to bookmarker.inc
 - added GIF version of all images. Test user-agent, and if known
   5.0 or later browser, show PNG image, otherwise show the GIF image.
 - added test in user page to make sure user ID matches [A-Za-Z0-9]
   
01/17/2000   Version 2.0.1
 - fixed bad create/insert sql in bookmarker.sql

01/16/2000   Version 2.0.0
 - split mailing lists. Now all discussion is on bookmarker, and
   bookmarker-announce will be announcements only.
 - track date that bookmark was added (new bookmark.added field)
 - fixed bug on mail-this-link that could cause site footer to be added multiple times
 - added graphics from tigert.gimp.org/ from gnome-stock
 - added .int as valid top-level domain (TLD) #114
 - added option on login to store auth as a permanent cookie
   on the local browser. this allows you to bypass the login
   on subsequent logins - ONLY set this when logging in via a
   secure machine. You'll need to update the user record before
   using this feature in order to set the perm_auth_cookie column
   for each user.

vbk.pl
#!/usr/local/bin/perl -w
#
# bookmarker verification perl script
# see vbk.sh for a handy shell script to run this
# via nohup
#
# make sure and update the database connection
# variables for your installation below.
#
use strict;

# use DBI.pm generic database interface
use DBI;

# use LWP for HTTP automation
use LWP::UserAgent;

# CGI.pm module
# set to -no_debug since we run from command line
# use standard HTML plus HTML 3 additions
# use Pretty formatted HTML so it is readable
# send errors to HTML output
use CGI qw(-no_debug :standard :html3);
use CGI::Pretty;
use CGI::Carp 'fatalsToBrowser';

my $ua = new LWP::UserAgent;
$ua->agent("bookmarker/1.0");
$ua->timeout(20);

my $driver = 'mysql';
my $database = 'db';
my $options = '';
my $user = 'user';
my $password = 'password';

my $title = 'Bookmarker URL Verification Report';

print start_html(-title=>$title,
                   -dtd=>1,
		   -BGCOLOR=>'white');

print h2 ($title);

my $dsn = "DBI:$driver:database=$database;$options";
 
my $dbh = DBI->connect($dsn, $user, $password) || die "Connect Failed: $DBI::errstr \n ";
 
my $sth = $dbh->prepare('select id, url, name from bookmark order by id')  || die "Select Failed: $DBI::errstr \n ";

$sth->execute || die "Execute Failed: $DBI::errstr \n ";

my ($id, $url, $name, $request, $response, $msg, $errs);

my @headings = ('ID','Bookmark', 'Response');
my @rows = th(\@headings);
     
$errs = 0;
while ( ($id, $url, $name) = $sth->fetchrow_array ) {

  $request = new HTTP::Request('HEAD',$url);
  $response = $ua->request($request); # fetch!
      
  if ($response->is_error) {
    $errs += 1;
    $msg = $response->code . ": " . $response->message;
    push(@rows, td([ a({-href=>"maintain.php3?id=$id"},$id)
                    ,a({-href=>$url},$name)
		    ,$msg]
		 ));
  }

}        	     
$sth->finish();
$dbh->disconnect;

if ( $errs > 0 ) {
  print strong ("$errs Failed URL Requests:");
  print table({-border=>'1',-width=>'90%'},
             Tr(\@rows)
             );
} else {
  print strong ('No Failed Requests!');
}

print end_html;

exit 0;


duser.sh
#!/bin/sh
# ---------------------------------------------------------------
# bookmarker
# A WWW based bookmark management, retrieval, and search tool.
# Copyright (C) 1998  Padraic Renaghan
# Licensed under terms of GNU General Public License
# (see http://www.renaghan.com/bookmarker/source/LICENSE)
# ---------------------------------------------------------------
# $Id: duser.sh,v 1.1 1999/10/29 21:04:27 prenagha Exp $
# ---------------------------------------------------------------
#
# duser
#       This shell script deletes a user and the user's data
#       from the database.
MYSQL=/usr/bin/mysql
DB=bookmarks
USER=test
PASS=test
TBLS="category rating subcategory search bookmark auth_user"

for tbl in $TBLS; do
  q="delete from $tbl where username = '$1';"
  echo "$q ..."
  echo $q | $MYSQL $DB -u $USER -p$PASS
done
exit 0


md5.js
/*
 *  md5.jvs 1.0b 27/06/96
 *
 * Javascript implementation of the RSA Data Security, Inc. MD5
 * Message-Digest Algorithm.
 *
 * Copyright (c) 1996 Henri Torgemane. All Rights Reserved.
 *
 * Permission to use, copy, modify, and distribute this software
 * and its documentation for any purposes and without
 * fee is hereby granted provided that this copyright notice
 * appears in all copies. 
 *
 * Of course, this soft is provided "as is" without express or implied
 * warranty of any kind.
 */
// $Id: md5.js,v 1.1 1999/04/17 19:45:27 prenagha Exp $


function array(n) {
  for(i=0;i<n;i++) this[i]=0;
  this.length=n;
}

/* Some basic logical functions had to be rewritten because of a bug in
 * Javascript.. Just try to compute 0xffffffff >> 4 with it..
 * Of course, these functions are slower than the original would be, but
 * at least, they work!
 */

function integer(n) { return n%(0xffffffff+1); }

function shr(a,b) {
  a=integer(a);
  b=integer(b);
  if (a-0x80000000>=0) {
    a=a%0x80000000;
    a>>=b;
    a+=0x40000000>>(b-1);
  } else
    a>>=b;
  return a;
}

function shl1(a) {
  a=a%0x80000000;
  if (a&0x40000000==0x40000000)
  {
    a-=0x40000000;  
    a*=2;
    a+=0x80000000;
  } else
    a*=2;
  return a;
}

function shl(a,b) {
  a=integer(a);
  b=integer(b);
  for (var i=0;i<b;i++) a=shl1(a);
  return a;
}

function and(a,b) {
  a=integer(a);
  b=integer(b);
  var t1=(a-0x80000000);
  var t2=(b-0x80000000);
  if (t1>=0) 
    if (t2>=0) 
      return ((t1&t2)+0x80000000);
    else
      return (t1&b);
  else
    if (t2>=0)
      return (a&t2);
    else
      return (a&b);  
}

function or(a,b) {
  a=integer(a);
  b=integer(b);
  var t1=(a-0x80000000);
  var t2=(b-0x80000000);
  if (t1>=0) 
    if (t2>=0) 
      return ((t1|t2)+0x80000000);
    else
      return ((t1|b)+0x80000000);
  else
    if (t2>=0)
      return ((a|t2)+0x80000000);
    else
      return (a|b);  
}

function xor(a,b) {
  a=integer(a);
  b=integer(b);
  var t1=(a-0x80000000);
  var t2=(b-0x80000000);
  if (t1>=0) 
    if (t2>=0) 
      return (t1^t2);
    else
      return ((t1^b)+0x80000000);
  else
    if (t2>=0)
      return ((a^t2)+0x80000000);
    else
      return (a^b);  
}

function not(a) {
  a=integer(a);
  return (0xffffffff-a);
}

/* Here begin the real algorithm */

    var state = new array(4); 
    var count = new array(2);
	count[0] = 0;
	count[1] = 0;                     
    var buffer = new array(64); 
    var transformBuffer = new array(16); 
    var digestBits = new array(16);

    var S11 = 7;
    var S12 = 12;
    var S13 = 17;
    var S14 = 22;
    var S21 = 5;
    var S22 = 9;
    var S23 = 14;
    var S24 = 20;
    var S31 = 4;
    var S32 = 11;
    var S33 = 16;
    var S34 = 23;
    var S41 = 6;
    var S42 = 10;
    var S43 = 15;
    var S44 = 21;

    function F(x,y,z) {
	return or(and(x,y),and(not(x),z));
    }

    function G(x,y,z) {
	return or(and(x,z),and(y,not(z)));
    }

    function H(x,y,z) {
	return xor(xor(x,y),z);
    }

    function I(x,y,z) {
	return xor(y ,or(x , not(z)));
    }

    function rotateLeft(a,n) {
	return or(shl(a, n),(shr(a,(32 - n))));
    }

    function FF(a,b,c,d,x,s,ac) {
        a = a+F(b, c, d) + x + ac;
	a = rotateLeft(a, s);
	a = a+b;
	return a;
    }

    function GG(a,b,c,d,x,s,ac) {
	a = a+G(b, c, d) +x + ac;
	a = rotateLeft(a, s);
	a = a+b;
	return a;
    }

    function HH(a,b,c,d,x,s,ac) {
	a = a+H(b, c, d) + x + ac;
	a = rotateLeft(a, s);
	a = a+b;
	return a;
    }

    function II(a,b,c,d,x,s,ac) {
	a = a+I(b, c, d) + x + ac;
	a = rotateLeft(a, s);
	a = a+b;
	return a;
    }

    function transform(buf,offset) { 
	var a=0, b=0, c=0, d=0; 
	var x = transformBuffer;
	
	a = state[0];
	b = state[1];
	c = state[2];
	d = state[3];
	
	for (i = 0; i < 16; i++) {
	    x[i] = and(buf[i*4+offset],0xff);
	    for (j = 1; j < 4; j++) {
		x[i]+=shl(and(buf[i*4+j+offset] ,0xff), j * 8);
	    }
	}

	/* Round 1 */
	a = FF ( a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
	d = FF ( d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
	c = FF ( c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
	b = FF ( b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
	a = FF ( a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
	d = FF ( d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
	c = FF ( c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
	b = FF ( b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
	a = FF ( a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
	d = FF ( d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
	c = FF ( c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
	b = FF ( b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
	a = FF ( a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
	d = FF ( d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
	c = FF ( c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
	b = FF ( b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

	/* Round 2 */
	a = GG ( a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
	d = GG ( d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
	c = GG ( c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
	b = GG ( b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
	a = GG ( a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
	d = GG ( d, a, b, c, x[10], S22,  0x2441453); /* 22 */
	c = GG ( c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
	b = GG ( b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
	a = GG ( a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
	d = GG ( d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
	c = GG ( c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
	b = GG ( b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
	a = GG ( a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
	d = GG ( d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
	c = GG ( c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
	b = GG ( b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

	/* Round 3 */
	a = HH ( a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
	d = HH ( d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
	c = HH ( c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
	b = HH ( b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
	a = HH ( a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
	d = HH ( d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
	c = HH ( c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
	b = HH ( b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
	a = HH ( a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
	d = HH ( d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
	c = HH ( c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
	b = HH ( b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
	a = HH ( a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
	d = HH ( d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
	c = HH ( c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
	b = HH ( b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

	/* Round 4 */
	a = II ( a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
	d = II ( d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
	c = II ( c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
	b = II ( b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
	a = II ( a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
	d = II ( d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
	c = II ( c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
	b = II ( b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
	a = II ( a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
	d = II ( d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
	c = II ( c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
	b = II ( b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
	a = II ( a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
	d = II ( d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
	c = II ( c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
	b = II ( b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

	state[0] +=a;
	state[1] +=b;
	state[2] +=c;
	state[3] +=d;

    }

    function init() {
	count[0]=count[1] = 0;
	state[0] = 0x67452301;
	state[1] = 0xefcdab89;
	state[2] = 0x98badcfe;
	state[3] = 0x10325476;
	for (i = 0; i < digestBits.length; i++)
	    digestBits[i] = 0;
    }

    function update(b) { 
	var index,i;
	
	index = and(shr(count[0],3) , 0x3f);
	if (count[0]<0xffffffff-7) 
	  count[0] += 8;
        else {
	  count[1]++;
	  count[0]-=0xffffffff+1;
          count[0]+=8;
        }
	buffer[index] = and(b,0xff);
	if (index  >= 63) {
	    transform(buffer, 0);
	}
    }

    function finish() {
	var bits = new array(8);
	var	padding; 
	var	i=0, index=0, padLen=0;

	for (i = 0; i < 4; i++) {
	    bits[i] = and(shr(count[0],(i * 8)), 0xff);
	}
        for (i = 0; i < 4; i++) {
	    bits[i+4]=and(shr(count[1],(i * 8)), 0xff);
	}
	index = and(shr(count[0], 3) ,0x3f);
	padLen = (index < 56) ? (56 - index) : (120 - index);
	padding = new array(64); 
	padding[0] = 0x80;
        for (i=0;i<padLen;i++)
	  update(padding[i]);
        for (i=0;i<8;i++) 
	  update(bits[i]);

	for (i = 0; i < 4; i++) {
	    for (j = 0; j < 4; j++) {
		digestBits[i*4+j] = and(shr(state[i], (j * 8)) , 0xff);
	    }
	} 
    }

/* End of the MD5 algorithm */

function hexa(n) {
 var hexa_h = "0123456789abcdef";
 var hexa_c=""; 
 var hexa_m=n;
 for (hexa_i=0;hexa_i<8;hexa_i++) {
   hexa_c=hexa_h.charAt(Math.abs(hexa_m)%16)+hexa_c;
   hexa_m=Math.floor(hexa_m/16);
 }
 return hexa_c;
}


var ascii="01234567890123456789012345678901" +
          " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ"+
          "[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";

function MD5(entree) 
{
 var l,s,k,ka,kb,kc,kd;

 init();
 for (k=0;k<entree.length;k++) {
   l=entree.charAt(k);
   update(ascii.lastIndexOf(l));
 }
 finish();
 ka=kb=kc=kd=0;
 for (i=0;i<4;i++) ka+=shl(digestBits[15-i], (i*8));
 for (i=4;i<8;i++) kb+=shl(digestBits[15-i], ((i-4)*8));
 for (i=8;i<12;i++) kc+=shl(digestBits[15-i], ((i-8)*8));
 for (i=12;i<16;i++) kd+=shl(digestBits[15-i], ((i-12)*8));
 s=hexa(kd)+hexa(kc)+hexa(kb)+hexa(ka);
 return s; 
}
